### PR TITLE
Compact logs based on oldest snapshot

### LINKED ISSUE #9627: Prevent compaction when backup is taken
While the backup is taken, the log segments needed for the backup must not be deleted. 

Possible solutions:
1. `LogCompactor ` allows "enabling" and "disabling" compaction, which backup manager can use. This was implemented in the prootype. While it works, there were some edge cases that are not handled. The edge-cases prevented compaction for a long time while there are multiple overlapping backups.
2. Instead of using the current snapshot, LogDeletionService looks for the oldest snapshot, and set the compactible index based on the oldest snapshot. Thus while a snapshot is "locked", the related segments are not deleted.

### PR DESCRIPTION
**Pull request by deepthidevaki**
## Description

For taking backups asynchronously, we have to keep the snapshot needed for the backup until the backup is completed. We should also not compact the log segments that are needed for the backup. To enable this we will now only compact the logs based on the oldest snapshot. The backup manager is expected to reserve the snapshot so that the snapshot store doesn't delete the snapshot. Thus even if there is a new snapshot, we won't compact the logs until the backup is completed.

## Related issues

closes #9627 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Review by lenaschoenburg — APPROVED** [review]
LGTM :+1: 

At first I was a bit concerned that the `LogDeletionService` could be triggering compaction based on an outdated list of available snapshots but then I realized that this shouldn't be an issue because snapshots are always starting off as being in the list of available snapshots and are only leaving this list if they aren't reserved.

File: broker/src/test/java/io/camunda/zeebe/broker/logstreams/LogDeletionServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    when(snapshotStore.getAvailableSnapshots()).thenReturn(actorFuture);
+
+    // when
+    service.onNewSnapshot(new TestPersistedSnapshot(11));
+
+    // then
+    verify(compactor, timeout(1000).times(1)).compactLog(5);
+  }
+
+  static class TestPersistedSnapshot implements PersistedSnapshot {
```
[the remark this answers is not in the document]
Comment (deepthidevaki) [reply broker/src/test/java/io/camunda/zeebe/broker/logstreams/LogDeletionServiceTest.java:71]: Let me try to rewrite it with mocks. I thought having a lot of mocks would not be so readable.

**Comment by deepthidevaki** [comment]
bors merge

**Comment by deepthidevaki** [comment]
bors merge

**Pull request merged**